src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,348 @@ public async Task TestRemoteWorkspaceCircularReferences()
+            bool syncWithRemoteServer, ImmutableArray<ImmutableArray<(string hintName, SourceText text)>> values)
+        {
+            var throwIfCalled = false;
+            ImmutableArray<(string hintName, SourceText text)> sourceTexts = default;
+            var generator = new CallbackGenerator(
+                onInit: _ => { },
+                onExecute: _ => { },
+                computeSourceTexts: () =>
+                {
+                    Contract.ThrowIfTrue(throwIfCalled);
+                    return sourceTexts;
+                });
+
+            using var localWorkspace = CreateWorkspace(syncWithRemoteServer ? Array.Empty<Type>() : new[] { typeof(NoSyncWorkspaceConfigurationService) });
+
+            DocumentId tempDocId;
+
+            // Keep this all in a nested scope so we don't accidentally access this data inside the loop below.  We only
+            // want to access the true workspace solution (which will be a fork of the solution we're producing here).
+            {
+                var projectId = ProjectId.CreateNewId();
+                var analyzerReference = new TestGeneratorReference(generator);
+                var project = localWorkspace.CurrentSolution
+                    .AddProject(ProjectInfo.Create(projectId, VersionStamp.Default, name: "Test", assemblyName: "Test", language: LanguageNames.CSharp))
+                    .GetRequiredProject(projectId)
+                    .AddAnalyzerReference(analyzerReference);
+                var tempDoc = project.AddDocument("X.cs", SourceText.From("// "));
+                tempDocId = tempDoc.Id;
+
+                Assert.True(localWorkspace.SetCurrentSolution(_ => tempDoc.Project.Solution, WorkspaceChangeKind.SolutionChanged));
+            }
+
+            using var client = await InProcRemoteHostClient.GetTestClientAsync(localWorkspace);
+            var remoteWorkspace = client.GetRemoteWorkspace();
+
+            for (var i = 0; i < values.Length; i++)
+            {
+                sourceTexts = values[i];
+
+                // make a change to the project to force a change between the local and oop solutions.
+
+                Assert.True(localWorkspace.SetCurrentSolution(s => s.WithDocumentText(tempDocId, SourceText.From("// " + i)), WorkspaceChangeKind.SolutionChanged));
+                await UpdatePrimaryWorkspace(client, localWorkspace.CurrentSolution);
+
+                var localProject = localWorkspace.CurrentSolution.Projects.Single();
+                var remoteProject = remoteWorkspace.CurrentSolution.Projects.Single();
+
+                // Run generators locally
+                throwIfCalled = false;
+                var localCompilation = await localProject.GetCompilationAsync();
+
+                // Now run them remotely.  This must not actually call into the generator since nothing has changed.
+                throwIfCalled = true;
+                var remoteCompilation = await remoteProject.GetCompilationAsync();
```

Comment:
Doesn't this need to be conditioned on syncWithRemoteServer? Becuse if we aren't syncing, then the local run would have stayed "local", and the remote run would run independently?